[orc8r][migration] Flip assoc directions between base_name, policy, subscriber

andreilee opened the pull request:
## Summary

Enforces the following assoc directions by swapping when necessary:
```
base_name -> policy_rule
base_name -> subscriber
subscriber -> policy_rule
```

## Test Plan

Ran locally with new copy of prod db.

Output below in a test where 7 edges were flipped
```
root@27307e374cf1:/var/opt/magma/bin# ./m012_policy_edge
I0824 00:22:34.173742     361 main.go:99] BEGIN MIGRATION
I0824 00:22:34.180711     361 main.go:110] RUNNING TRANSACTION
I0824 00:22:34.185308     361 main.go:117] Swapping 4 assoc directions to enforce policy -> base_name
I0824 00:22:34.185397     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [08366cae-668f-4836-bf78-6327f32e3aff 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 08366cae-668f-4836-bf78-6327f32e3aff]
I0824 00:22:34.186951     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [4a6992d3-6c1c-426b-ac21-22c2514158e3 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 4a6992d3-6c1c-426b-ac21-22c2514158e3]
I0824 00:22:34.187950     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [e07e2de4-3994-4657-908f-527730ad1911 8a035a60-b43a-498f-a4cf-436d74564deb 8a035a60-b43a-498f-a4cf-436d74564deb e07e2de4-3994-4657-908f-527730ad1911]
I0824 00:22:34.189200     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [a91ec1fc-b105-44d6-8a01-71efdca886de 5e679f8a-b8b1-4231-892c-7c407efb068b 5e679f8a-b8b1-4231-892c-7c407efb068b a91ec1fc-b105-44d6-8a01-71efdca886de]
I0824 00:22:34.197652     361 main.go:128] Swapping 1 assoc directions to enforce subscriber -> base_name
I0824 00:22:34.197814     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [720fe3ac-4ed8-42f7-a5bb-e8415fce60f7 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 720fe3ac-4ed8-42f7-a5bb-e8415fce60f7]
I0824 00:22:34.203577     361 main.go:139] Swapping 2 assoc directions to enforce subscriber -> policy
I0824 00:22:34.203649     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [720fe3ac-4ed8-42f7-a5bb-e8415fce60f7 08366cae-668f-4836-bf78-6327f32e3aff 08366cae-668f-4836-bf78-6327f32e3aff 720fe3ac-4ed8-42f7-a5bb-e8415fce60f7]
I0824 00:22:34.204566     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [720fe3ac-4ed8-42f7-a5bb-e8415fce60f7 4a6992d3-6c1c-426b-ac21-22c2514158e3 4a6992d3-6c1c-426b-ac21-22c2514158e3 720fe3ac-4ed8-42f7-a5bb-e8415fce60f7]
I0824 00:22:34.206642     361 main.go:156] SUCCESS
I0824 00:22:34.206685     361 main.go:157] END MIGRATION
```

Verifying idempotency by running again afterwards:
```
root@27307e374cf1:/var/opt/magma/bin# ./m012_policy_edge
I0824 00:22:34.173742     361 main.go:99] BEGIN MIGRATION
I0824 00:22:34.180711     361 main.go:110] RUNNING TRANSACTION
I0824 00:22:34.185308     361 main.go:117] Swapping 4 assoc directions to enforce policy -> base_name
I0824 00:22:34.185397     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [08366cae-668f-4836-bf78-6327f32e3aff 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 08366cae-668f-4836-bf78-6327f32e3aff]
I0824 00:22:34.186951     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [4a6992d3-6c1c-426b-ac21-22c2514158e3 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 4a6992d3-6c1c-426b-ac21-22c2514158e3]
I0824 00:22:34.187950     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [e07e2de4-3994-4657-908f-527730ad1911 8a035a60-b43a-498f-a4cf-436d74564deb 8a035a60-b43a-498f-a4cf-436d74564deb e07e2de4-3994-4657-908f-527730ad1911]
I0824 00:22:34.189200     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [a91ec1fc-b105-44d6-8a01-71efdca886de 5e679f8a-b8b1-4231-892c-7c407efb068b 5e679f8a-b8b1-4231-892c-7c407efb068b a91ec1fc-b105-44d6-8a01-71efdca886de]
I0824 00:22:34.197652     361 main.go:128] Swapping 1 assoc directions to enforce subscriber -> base_name
I0824 00:22:34.197814     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [720fe3ac-4ed8-42f7-a5bb-e8415fce60f7 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 720fe3ac-4ed8-42f7-a5bb-e8415fce60f7]
I0824 00:22:34.203577     361 main.go:139] Swapping 2 assoc directions to enforce subscriber -> policy
I0824 00:22:34.203649     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [720fe3ac-4ed8-42f7-a5bb-e8415fce60f7 08366cae-668f-4836-bf78-6327f32e3aff 08366cae-668f-4836-bf78-6327f32e3aff 720fe3ac-4ed8-42f7-a5bb-e8415fce60f7]
I0824 00:22:34.204566     361 main.go:251] [RUN] UPDATE cfg_assocs SET from_pk = $1, to_pk = $2 WHERE from_pk = $3 AND to_pk = $4 [720fe3ac-4ed8-42f7-a5bb-e8415fce60f7 4a6992d3-6c1c-426b-ac21-22c2514158e3 4a6992d3-6c1c-426b-ac21-22c2514158e3 720fe3ac-4ed8-42f7-a5bb-e8415fce60f7]
I0824 00:22:34.206642     361 main.go:156] SUCCESS
I0824 00:22:34.206685     361 main.go:157] END MIGRATION
root@27307e374cf1:/var/opt/magma/bin# ./m012_policy_edge
I0824 00:25:10.932655     369 main.go:99] BEGIN MIGRATION
I0824 00:25:10.939827     369 main.go:110] RUNNING TRANSACTION
I0824 00:25:10.942988     369 main.go:117] Swapping 0 assoc directions to enforce policy -> base_name
I0824 00:25:10.949042     369 main.go:128] Swapping 0 assoc directions to enforce subscriber -> base_name
I0824 00:25:10.954567     369 main.go:139] Swapping 0 assoc directions to enforce subscriber -> policy
I0824 00:25:10.955240     369 main.go:156] SUCCESS
I0824 00:25:10.955290     369 main.go:157] END MIGRATION
```

Output below querying the db for edge swaps, before migration
```
prod-1-aug-21=# -- This finds all the assocs we need to flip -> really only need to select from_pk,to_pk
prod-1-aug-21=# SELECT a.key,a.pk,a.type, b.key,b.pk,b.type
prod-1-aug-21-# FROM cfg_entities AS a
prod-1-aug-21-# JOIN cfg_assocs ON a.pk=from_pk
prod-1-aug-21-# JOIN cfg_entities AS b ON to_pk=b.pk
prod-1-aug-21-# WHERE
prod-1-aug-21-#     (a.type='base_name' AND b.type='policy') OR
prod-1-aug-21-#     (a.type='policy' AND b.type='subscriber') OR
prod-1-aug-21-#     (a.type='base_name' AND b.type='subscriber');
     key     |                  pk                  |   type    |         key         |                  pk                  |    type
-------------+--------------------------------------+-----------+---------------------+--------------------------------------+------------
 test_rule   | 08366cae-668f-4836-bf78-6327f32e3aff | policy    | IMSI001010000000022 | 720fe3ac-4ed8-42f7-a5bb-e8415fce60f7 | subscriber
 test_rule_2 | 4a6992d3-6c1c-426b-ac21-22c2514158e3 | policy    | IMSI001010000000022 | 720fe3ac-4ed8-42f7-a5bb-e8415fce60f7 | subscriber
 base_1      | 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 | base_name | test_rule           | 08366cae-668f-4836-bf78-6327f32e3aff | policy
 base_1      | 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 | base_name | test_rule_2         | 4a6992d3-6c1c-426b-ac21-22c2514158e3 | policy
 base_1      | 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 | base_name | IMSI001010000000022 | 720fe3ac-4ed8-42f7-a5bb-e8415fce60f7 | subscriber
 freepass    | 5e679f8a-b8b1-4231-892c-7c407efb068b | base_name | free_dns            | a91ec1fc-b105-44d6-8a01-71efdca886de | policy
 freepass    | 8a035a60-b43a-498f-a4cf-436d74564deb | base_name | free_dns            | e07e2de4-3994-4657-908f-527730ad1911 | policy
(7 rows)

prod-1-aug-21=# -- This finds all the assocs we have flipped already
prod-1-aug-21=# SELECT a.key,a.pk,a.type, b.key,b.pk,b.type
prod-1-aug-21-# FROM cfg_entities AS a
prod-1-aug-21-# JOIN cfg_assocs ON a.pk=from_pk
prod-1-aug-21-# JOIN cfg_entities AS b ON to_pk=b.pk
prod-1-aug-21-# WHERE
prod-1-aug-21-#     (a.type='policy' AND b.type='base_name') OR
prod-1-aug-21-#     (a.type='subscriber' AND b.type='policy') OR
prod-1-aug-21-#     (a.type='subscriber' AND b.type='base_name');
 key | pk | type | key | pk | type
-----+----+------+-----+----+------
(0 rows)
```

And output after db migration:
```
prod-1-aug-21=# -- This finds all the assocs we need to flip -> really only need to select from_pk,to_pk
prod-1-aug-21=# SELECT a.key,a.pk,a.type, b.key,b.pk,b.type
prod-1-aug-21-# FROM cfg_entities AS a
prod-1-aug-21-# JOIN cfg_assocs ON a.pk=from_pk
prod-1-aug-21-# JOIN cfg_entities AS b ON to_pk=b.pk
prod-1-aug-21-# WHERE
prod-1-aug-21-#     (a.type='base_name' AND b.type='policy') OR
prod-1-aug-21-#     (a.type='policy' AND b.type='subscriber') OR
prod-1-aug-21-#     (a.type='base_name' AND b.type='subscriber');
 key | pk | type | key | pk | type
-----+----+------+-----+----+------
(0 rows)

prod-1-aug-21=# -- This finds all the assocs we have flipped already
prod-1-aug-21=# SELECT a.key,a.pk,a.type, b.key,b.pk,b.type
prod-1-aug-21-# FROM cfg_entities AS a
prod-1-aug-21-# JOIN cfg_assocs ON a.pk=from_pk
prod-1-aug-21-# JOIN cfg_entities AS b ON to_pk=b.pk
prod-1-aug-21-# WHERE
prod-1-aug-21-#     (a.type='policy' AND b.type='base_name') OR
prod-1-aug-21-#     (a.type='subscriber' AND b.type='policy') OR
prod-1-aug-21-#     (a.type='subscriber' AND b.type='base_name');
         key         |                  pk                  |    type    |     key     |                  pk                  |   type
---------------------+--------------------------------------+------------+-------------+--------------------------------------+-----------
 free_dns            | e07e2de4-3994-4657-908f-527730ad1911 | policy     | freepass    | 8a035a60-b43a-498f-a4cf-436d74564deb | base_name
 free_dns            | a91ec1fc-b105-44d6-8a01-71efdca886de | policy     | freepass    | 5e679f8a-b8b1-4231-892c-7c407efb068b | base_name
 test_rule           | 08366cae-668f-4836-bf78-6327f32e3aff | policy     | base_1      | 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 | base_name
 test_rule_2         | 4a6992d3-6c1c-426b-ac21-22c2514158e3 | policy     | base_1      | 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 | base_name
 IMSI001010000000022 | 720fe3ac-4ed8-42f7-a5bb-e8415fce60f7 | subscriber | base_1      | 8e9ece2d-74e4-4f24-a679-5e2a72edd7b2 | base_name
 IMSI001010000000022 | 720fe3ac-4ed8-42f7-a5bb-e8415fce60f7 | subscriber | test_rule   | 08366cae-668f-4836-bf78-6327f32e3aff | policy
 IMSI001010000000022 | 720fe3ac-4ed8-42f7-a5bb-e8415fce60f7 | subscriber | test_rule_2 | 4a6992d3-6c1c-426b-ac21-22c2514158e3 | policy
(7 rows)
```

## Additional Information

- [x] This change is backwards-breaking

To complete the required change, run the m012_policy_edge data migration.

andreilee committed:
Merge pull request #1 from magma/master

Update forked repo

andreilee committed:
Merge remote-tracking branch 'upstream/master'

andreilee committed:
[nms] Update flow version to 0.130.0

Signed-off-by: Andrei Lee <andreilee@fb.com>

andreilee committed:
Merge remote-tracking branch 'upstream/master'

andreilee committed:
Merge remote-tracking branch 'upstream/master'

andreilee committed:
[nms] Do not update flow version

Signed-off-by: Andrei Lee <andreilee@fb.com>

andreilee committed:
Merge remote-tracking branch 'upstream/master'

andreilee committed:
[orc8r][migration_required] Enforce assoc directions between base_name, policy, subscriber

Signed-off-by: Andrei Lee <andreilee@fb.com>

andreilee committed:
Merge remote-tracking branch 'upstream/master'

hcgatewood reviewed (changes requested):
Couple things

1. Do the read in a single call (see below)
2. I don't know of a good general way to do the writes in a single call, so what you have looks good
3. Pre-validation, create some policy->subscriber and base_name->subscriber edges through the rest api, as the prod db I sent you doesn't have any

```sql
# Still create this statement using sqorc
# Also paste this output before/after the migration as the test plan for this PR
SELECT a.key,a.pk,a.type, b.key,b.pk,b.type  # really only need to select from_pk,to_pk
FROM cfg_entities AS a
JOIN cfg_assocs ON a.pk=from_pk
JOIN cfg_entities AS b ON to_pk=b.pk
WHERE
    (a.type='base_name' AND b.type='policy') OR
    (a.type='policy' AND b.type='subscriber') OR
    (a.type='base_name' AND b.type='subscriber') ;
```

andreilee commented:
> Couple things
> 
> 1. Do the read in a single call (see below)
> 2. I don't know of a good general way to do the writes in a single call, so what you have looks good
> 3. Pre-validation, create some policy->subscriber and base_name->subscriber edges through the rest api, as the prod db I sent you doesn't have any
> 
> ```sql
> # Still create this statement using sqorc
> # Also paste this output before/after the migration as the test plan for this PR
> SELECT a.key,a.pk,a.type, b.key,b.pk,b.type  # really only need to select from_pk,to_pk
> FROM cfg_entities AS a
> JOIN cfg_assocs ON a.pk=from_pk
> JOIN cfg_entities AS b ON to_pk=b.pk
> WHERE
>     (a.type='base_name' AND b.type='policy') OR
>     (a.type='policy' AND b.type='subscriber') OR
>     (a.type='base_name' AND b.type='subscriber') ;
> ```

1. Let's just go with what's in the PR since there aren't many edges to flip. We can extract an assoc flipping method for later migrations after.
2. sgtm
3. Added data, modified test plan for updated data

hcgatewood reviewed (approved):
I still prefer the simplified select but we'll merge since we're crunched for time

Marking as non-breaking change since this diff doesn't actually break any functionality [0]. Will mark the soon-to-be-opened breaking PRs as breaking changes and point them to this PR

[0] Normally we want to include both the break and the migration in the same diff, but again since we're on a deadline we'll go with this
